lib/file-source/src/checkpointer.rs

```diff
@@ -264,20 +293,35 @@ impl Checkpointer {
         // matter anymore.
         self.checkpoints.remove_expired();
 
-        // Write the new checkpoints to a tmp file and flush it fully to
-        // disk. If vector dies anywhere during this section, the existing
-        // stable file will still be in its current valid state and we'll be
-        // able to recover.
-        let mut f = io::BufWriter::new(fs::File::create(&self.tmp_file_path)?);
-        serde_json::to_writer(&mut f, &self.checkpoints.get_state())?;
-        f.into_inner()?.sync_all()?;
-
-        // Once the temp file is fully flushed, rename the tmp file to replace
-        // the previous stable file. This is an atomic operation on POSIX
-        // systems (and the stdlib claims to provide equivalent behavior on
-        // Windows), which should prevent scenarios where we don't have at least
-        // one full valid file to recover from.
-        fs::rename(&self.tmp_file_path, &self.stable_file_path)?;
+        let current = self.checkpoints.get_state();
+
+        // Fetch last written state.
+        let mut last = match self.last.lock() {
+            Ok(guard) => guard,
+            Err(mut poisoned) => {
+                // Remove possibly invalid state.
+                **poisoned.get_mut() = None;
+                poisoned.into_inner()
+            }
```

Comment:
If the lock is poisoned we should just panic. There's no valid path where that should happen and it'd be better to shut down than continue.